### ISSUE TITLE
Datepicker - form is invalid with Long or String values

### ISSUE THREAD

**Issue by apandichi** [angular-ui/bootstrap#3403]
After setting a Long or String value to the ng-model of a datepicker, the form $invalid value is set to 'true'. Please see this Plunkr for a demo of this behaviour: http://plnkr.co/edit/L4MpJxYB0OYkn80FB9hw

I'm not sure if this is the intended behaviour, but it seems rather odd that the input text value is formatted correctly to a Date, but the form is set to invalid.
Any thoughts? Thanks!

P.S. It's funny that the only time when the form seems to be valid is when the 'dd.MM.yyyy' formatting is not applied, when setting the Date value.

**Comment by RobJacobs** [comment]
I think this issue is related to [this](https://github.com/angular-ui/bootstrap/pull/3322/files) pull request.

**Comment by wesleycho** [comment]
I believe this issue is currently fixed in master - see the forked version of that Plunker based on current `master` [here](http://plnkr.co/edit/U3nuipb1Z0aLhLCRS536?p=info).

Let us know if this is still an issue.

**Comment by apandichi** [comment]
Hi @wesleycho, thanks for your input.

In your new plunkr, the date input is no longer formatted when String and Long values are set. Also, the  openStartDate($event) button on the right does not work anymore. When running the plunkr, the JS console is throwing an error: "TypeError: Cannot read property 'initDate' of undefined"

Is this expected behaviour? I see that ui-bootstrap.js version is 0.13.1-SNAPSHOT, so I'm guessing not all issues have been ironed out?

In conclusion, what I'd like to know is if I can expect the datepicker to accept the String and Long values as valid values, and if the validity will be reflected by the form $invalid value.

Thanks a lot!

**Comment by wesleycho** [comment]
Hmm, so it looks like

**Comment by wesleycho** [comment]
I will take a look at that today then.

**Comment by wesleycho** [comment]
I updated the Plunker and removed the datepicker-options attribute there, since it was set to `undefined`. It appears that the formatting works, but the validation is not right when the number is set to a string representation. This would be buggy behavior here still and is a legitimate issue.

**Comment by chrisirhc** [comment]
I'm thinking we should consider dropping support for any input type that can't be parsed using the date format available described in https://docs.angularjs.org/api/ng/filter/date .

The reason for this is that this violates the idea of the date format. The format can't be used for parsing and if it's used for output, it makes the output/input asymmetric (input is number, output is string). That's quite confusing because then it's hard to define what's the output of the directive. It also violates the principle of least surprise (you put a number in the model, and applying a directive makes it into a string after modification of the field, try that on the [Plunker here](http://plnkr.co/edit/HVwMYo8TyN5ag5y6IrRR?p=preview)).

Adding a custom format string for unix timestamps is one way to do this. However, I think Unix timestamp support should be handled separately as a separate directive, via a parser/formatter.

**Comment by wesleycho** [comment]
As we dropped date support that is not parsed by the dateparser, this ticket is now resolved as far as this particular request goes.

On the TODO list is to allow users to use their own model for dates, which is encapsulated by #4093.
